src/lib/Imp.hs

```diff
@@ -212,40 +234,60 @@ intToIndex ty@(TC con) i = case con of
   RecType r -> do
     strides <- getStrides $ fmap (\t->(t,t)) r
     liftM RecVal $
-      flip evalStateT i $ forM strides $ \(ty', stride) -> do
+      flip evalStateT i $ forM strides $ \(ty', _, stride) -> do
         i' <- get
         iCur  <- lift $ impDiv i' stride
         iRest <- lift $ impRem i' stride
         put iRest
         lift $ intToIndex ty' iCur
+  SumType (l, r) -> do
+    ls <- indexSetSize l
+    isLeft <- impCmp Less i ls
+    li <- intToIndex l i
+    ri <- intToIndex r =<< impSub i ls
+    return $ Con $ SumCon (toScalarAtom BoolTy isLeft) li ri
   _ -> error $ "Unexpected type " ++ pprint con
   where
     iAsIdx = return $ Con $ AsIdx ty $ impExprToAtom i
 intToIndex ty _ = error $ "Unexpected type " ++ pprint ty
 
-indexToInt :: Atom -> ImpM IExpr
-indexToInt idx | getType idx == TC (BaseType BoolType) = emitUnOp BoolToInt =<< fromScalarAtom idx
-indexToInt idx = case idx of
-  RecVal r -> do
-    rWithStrides <- getStrides $ fmap (\x -> (x, getType x)) r
-    foldrM f (IIntVal 0) rWithStrides
-    where
-      f :: (Atom, IExpr) -> IExpr -> ImpM IExpr
-      f (i, stride) cumIdx = do
-        i' <- indexToInt i
-        iDelta  <- impMul i' stride
-        impAdd cumIdx iDelta
-  _ -> fromScalarAtom idx
-
-getStrides :: Traversable f => f (a, Type) -> ImpM (f (a, IExpr))
+indexToInt :: Type -> Atom -> ImpM IExpr
+indexToInt ty idx = case ty of
```

Comment:
I have refactored this function to also get the type as an argument, because it's always readily available at the call sites and otherwise we would have to retrieve it ourselves in here...